File: src/main/java/tk/darkdustry/bot/components/SchematicBatch.java
package tk.darkdustry.bot.components;

import arc.graphics.Texture;
import arc.graphics.g2d.*;
import arc.graphics.g2d.TextureAtlas.AtlasRegion;
import arc.math.Mathf;

import java.awt.geom.AffineTransform;

import static tk.darkdustry.bot.Vars.*;

public class SchematicBatch extends SpriteBatch {

    public SchematicBatch() {
        super(0);
    }

    @Override
    protected void draw(TextureRegion region, float x, float y, float originX, float originY, float width, float height, float rotation) {
        x += 4;
        y += 4;
        x *= 4;
        y *= 4;

        var transform = new AffineTransform();
        transform.translate(x, currentImage.getHeight() - height * 4f - y);
        transform.rotate(-rotation * Mathf.degRad, originX * 4, originY * 4);

        currentGraphics.setTransform(transform);
        currentGraphics.drawImage(regions.get(((AtlasRegion) region).name), 0, 0, (int) width * 4, (int) height * 4, null);
    }

    @Override
    protected void draw(Texture texture, float[] spriteVertices, int offset, int count) {}
}

File: src/main/kotlin/tk/darkdustry/bot/components/SchematicBatch.kt
package tk.darkdustry.bot.components

import arc.graphics.Texture
import arc.graphics.g2d.SpriteBatch
import arc.graphics.g2d.TextureRegion
import arc.math.Mathf.degRad
import arc.util.Tmp
import tk.darkdustry.bot.currentGraphics
import tk.darkdustry.bot.currentImage
import java.awt.geom.AffineTransform


class SchematicBatch : SpriteBatch(0) {
    override fun draw(
        region: TextureRegion,
        x: Float,
        y: Float,
        originX: Float,
        originY: Float,
        width: Float,
        height: Float,
        rotation: Float
    ) {
        val tx = x * 4 + 16
        val ty = currentImage!!.height - height - y * 4 - 16

        val at = AffineTransform()
        at.translate(tx.toDouble(), ty.toDouble())

        at.rotate(-rotation.toDouble() * degRad, originX.toDouble() * 4, originY.toDouble() * 4)
        currentGraphics!!.transform = at

        val image = (region as ImageRegion).image
        for (x in 0 until image.width)
            for (y in 0 until image.height)
                image.setRGB(x, y, Tmp.c1.argb8888(image.getRGB(x, y)).mul(color).argb8888())

        currentGraphics!!.drawImage(image, 0, 0, (width * 4).toInt(), (height * 4).toInt(), null)
    }

    override fun draw(texture: Texture?, spriteVertices: FloatArray?, offset: Int, count: Int) {}
}